docs/developers/library_examples.md

```diff
@@ -0,0 +1,971 @@
+
+    if (!out) {
+        // Handle error.
+    }
+
+    std::array<char, 4_Mb> buffer{}; // Buffer full of data.
+
+    // This is the fastest way to write data into iRODS via the new stream API.
+    // Bytes written this way are stored directly in the buffer as is.
+    out.write(buffer.data(), buffer.size());
+
+    // This will also write data into the data object. This is slower than the previous method
+    // because stream operators format data.
+    out << "Here is some more data ...\n";
+}
+
+void read_from_data_object()
+{
+    namespace io = irods::experimental::io;
+
+    auto conn = // Our iRODS connection.
+
+    // See function above for information about this type.
+    io::client::default_transport xport{conn};
+
+    // Here, we are creating a new input stream for reading. 
+    io::idstream in{xport, "/path/to/data_object"};
+
+    if (!in) {
+        // Handle error.
+    }
+
+    std::array<char, 4_Mb> buffer{}; // Buffer to hold data.
+
+    // This is the fastest way to write data into iRODS via the new stream API.
+    // Bytes read this way are stored directly in the buffer as is.
+    in.read(buffer.data(), buffer.size());
+
+    // Read a single character sequence into "word".
+    // This assumes the input stream contains a sequence of printable characters.
+    std::string word;
+    in >> word;
+
+    std::string line;
+    while (std::getline(in, line)) {
+        // Read every line of the input stream until eof.
+    }
+}
+```
+
+### key_value_proxy
+
+Available Since: v4.2.8
+
+Demonstrates how to use `irods::key_value_proxy` to manipulate the C data type, `KeyValPair`.
+
+```c++
+#include <irods/key_value_proxy.hpp>
+
+#include <irods/dataObjOpen.h>
+#include <irods/stringOpr.h>
+
+#include <string>
+
+void manipulating_keyValuePair_t()
+{
+    // Let's open a replica for writing using the iRODS C API.
+    dataObjInp_t input{};
+    input.createMode = 0600;
+    input.openFlags = O_WRONLY;
+    rstrcpy(input.objPath, "/tempZone/home/rods/foo", MAX_NAME_LEN);
+
+    // Now, we need to set the options that target a specific replica.
+    // Normally we'd use the "keyValuePair_t" family of functions to add/remove
+    // options. Instead, we'll use the "key_value_proxy" class.
+    irods::experimental::key_value_proxy kvp{input.condInput};
+
+    // This proxy object does not own the underlying keyValuePair_t. It simply provides
+    // a map-like interface to manipulate and inspect it.
+
+    // Let's target a specific replica of this data object.
+    // This adds the "REPL_NUM_KW" key to the underlying keyValuePair_t and sets its
+    // value to the string "2". The string is copied into the object.
+    kvp[REPL_NUM_KW] = "2";
+
+    // This proxy type also supports checking if the kevValuePair_t contains a specific key.
+    if (kvp.contains(RESC_NAME_KW)) {
+        // Do something ...
+    }
+
+    // Extracting a value is easy too.
+    const std::string value = kvp[RESC_HIER_STR_KW];
```

Comment:
As a best practice, maybe we should use `at` here? Or mention it?

Reply:
Absolutely.

Reply:
Done.